### PR TITLE
Always close the engine

### LINKED ISSUE #623: Process leaks eventually resulting in OOM
Since setting up lichess-bot, I've noticed that my server eventually becomes unresponsive, forcing me to hard reset, with OOM errors spammed into my system logs. Monitoring the system using top, I notice that lichess-bot will occasionally spawn but fail to kill instances of my chess engine, often in pairs with sequential PIDs which makes me think it's spawning multiple engines at once.

I'm not entirely sure what kind of situation causes this, if anything in particular, since I've failed to catch it happening in real-time to see what lichess-bot prints out to the console, but I can update this with additional info if I finally manage to catch it. I'll probably start piping the output through tee to log everything.

If it's relevant, I'm running lichess-bot on Python 3.9.14 on DragonflyBSD.

### PR DESCRIPTION
**Pull request by AttackingOrDefending**
Makes sure that the engine is closed. The only error this won't catch is if it happens inside `__init__` in `UCIEngine`, `XBoardEngine` or `MinimalEngine`.

fixes #623

### CODE REVIEW

**Comment by MarkZH** [comment]
There's a way to do this without increasing the complexity of `play_game()` using a [context manager decorator](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) on `create_engine()`. The the `create_engine()` function would look like this:
```python
# engine_wrapper.py

from contextlib import contextmanager

@contextmanager
def create_engine(config):

    # ... parse options

    engine = Engine(commands, options, stderr, cfg.draw_or_resign, cwd=cfg.working_dir)
    try:
        yield engine
    finally:
        engine.stop()
        engine.ping()
        engine.quit()
```
Then, `create_engine()` and succeeding `try:` in `play_game()` is replaced with:
```python
# lichess-bot.py

    with engine_wrapper.create_engine(config) as engine:
        # ... the rest
```
The `finally:` block associated with that `try:` can be deleted since it will be handled by the `finally:` in `create_engine()`.

File: model.py [hunk 1]
```diff
@@ -98,7 +98,8 @@ def challenger_full_name(self):
         return f'{self.challenger_title or ""} {self.challenger_name}'.strip()
 
     def __str__(self):
-        return f"{self.perf_name} {self.mode()} challenge from {self.challenger_full_name()}({self.challenger_rating}) ({self.id})"
+        return f"{self.perf_name} {self.mode()} challenge from {self.challenger_full_name()}({self.challenger_rating})" \
+               f" ({self.id})"
```
Comment (MarkZH) [inline model.py:102]: Putting parentheses around the multi-line string is less fragile than escaping the newline. Accidentally putting space characters after the backslash will cause errors well after the program starts.

**Comment by MarkZH** [comment]
Let me know once this is ready to merge.

**Comment by AttackingOrDefending** [comment]
Ready.

**Review by MarkZH — APPROVED** [review]

**Pull request merged**